service/efsPort.py

```diff
@@ -0,0 +1,630 @@
+        sFit.removeModule(fitID, mwdPosition)
+        sFit.recalc(fit)
+        fit = eos.db.getFit(fitID)
+        return mwdPropSpeed
+
+    @staticmethod
+    def getPropData(fit, sFit):
+        fitID = fit.ID
+        propGroupId = getGroup("Propulsion Module").ID
+        propMods = filter(lambda mod: mod.item and mod.item.groupID == propGroupId, fit.modules)
+        activePropWBloomFilter = lambda mod: mod.state > 0 and "signatureRadiusBonus" in mod.item.attributes
+        propWithBloom = next(filter(activePropWBloomFilter, propMods), None)
+        if propWithBloom is not None:
+            oldPropState = propWithBloom.state
+            propWithBloom.state = 0
+            sFit.recalc(fit)
+            fit = eos.db.getFit(fitID)
+            sp = fit.maxSpeed
+            sig = fit.ship.itemModifiedAttributes["signatureRadius"]
+            propWithBloom.state = oldPropState
+            sFit.recalc(fit)
+            fit = eos.db.getFit(fitID)
+            return {"usingMWD": True, "unpropedSpeed": sp, "unpropedSig": sig}
+        return {
+            "usingMWD": False,
+            "unpropedSpeed": fit.maxSpeed,
+            "unpropedSig": fit.ship.itemModifiedAttributes["signatureRadius"]
+        }
+
+    @staticmethod
+    def getOutgoingProjectionData(fit):
+        # This is a subset of module groups capable of projection and a superset of those currently used by efs
+        modGroupNames = [
+            "Remote Shield Booster", "Warp Scrambler", "Stasis Web", "Remote Capacitor Transmitter",
+            "Energy Nosferatu", "Energy Neutralizer", "Burst Jammer", "ECM", "Sensor Dampener",
+            "Weapon Disruptor", "Remote Armor Repairer", "Target Painter", "Remote Hull Repairer",
+            "Burst Projectors", "Warp Disrupt Field Generator", "Armor Resistance Shift Hardener",
+            "Target Breaker", "Micro Jump Drive", "Ship Modifiers", "Stasis Grappler",
+            "Ancillary Remote Shield Booster", "Ancillary Remote Armor Repairer",
+            "Titan Phenomena Generator", "Non-Repeating Hardeners"
+        ]
+        modGroupIds = list(map(lambda s: getGroup(s).ID, modGroupNames))
+        modGroupData = dict(map(lambda name, gid: (name, {"name": name, "id": gid}),
+                                    modGroupNames, modGroupIds))
+        projectedMods = list(filter(lambda mod: mod.item and mod.item.groupID in modGroupIds, fit.modules))
+        projections = []
+        for mod in projectedMods:
+            stats = {}
+            if mod.item.groupID in [modGroupData["Stasis Web"]["id"], modGroupData["Stasis Grappler"]["id"]]:
+                stats["type"] = "Stasis Web"
+                stats["optimal"] = mod.itemModifiedAttributes["maxRange"]
+                EfsPort.attrDirectMap(["duration", "speedFactor"], stats, mod)
+            elif mod.item.groupID == modGroupData["Weapon Disruptor"]["id"]:
+                stats["type"] = "Weapon Disruptor"
```

Comment:
You might be able to use `mod.item.group.name` here to avoid hardcoded values, depending on how it applies to your application :)

Reply:
Good point, that cleans the whole thing up and massively simplifies getting the projectedMods list itself.